[1기 김문희] 지하철 노선도 SPA

moonheekim0118 opened the pull request:
## 데모
- https://quirky-archimedes-88d9e7.netlify.app

> 배포를했긴 했는데 비밀번호 암호화 없이 로컬스토리지에 넣어버려서 그런지 크롬에서 사기성 웹사이트라고 감지하네요ㅋㅋ

### 해결 못한 점
- 구간 기능 : 시간이 부족해서 구간 관련 기능은 구현못했습니다 😭 
- 새로고침 시 라우팅 : 해시가 아니라 / 로 라우팅을 하다보니 새로고침을 하니까 서버에 요청을 보내버려서 404가 뜹니다.

그 외에도, 처음에 페이지 들어가면 바로 그냥 로그인 페이지로 리다이렉트 시켜버렸는데 .. 이렇게 하는게 맞는건지 헷갈리네요..그리고 Station 등록하는 페이지를 그냥 인덱스 (/)페이지로 등록해놓았는데..이것도 요구사항에 맞는건지 헷갈리네욤..<br/>

<br/>

바닐라 자바스크립트로 라우팅 처음 해보는데 정말 어려웠던만큼 많은걸 배워갑니다 😃 

## 🎯 step1

### Webpack을 통한 번들링

- [x] Webpack에서 babel을 설정한다.
- [x] 기타 플러그인 및 설정은 본인이 필요에 따라 추가한다.

### 라우팅 기능

- [x] Browser History Api를 이용하여 SPA처럼 라우팅을 적용한다.

### 지하철역 관리 기능

- [x] 지하철역을 등록할 수 있다.
  - [x] 지하철역은 2글자 이상이어야 한다.
  - [x] 엔터키 또는 마우스 클릭으로 역을 추가할 수 있어야 한다.
  - [x] 지하철역의 이름은 최대 20자 이하이다.
  - [x] 중복된 지하철역은 추가할 수 없다.
- [x] 지하철역의 이름을 수정할 수 있다.
  - [x] 수정하는 UI는 modal을 이용하여 수정할 수 있게 한다.
- [x] 지하철역을 삭제할 수 있다.
  - [x] 삭제 시 confirm을 이용하여 한 번 더 유저가 확인할 수 있어야 한다.
  - [x] 이미 노선에 등록된 역인 경우 삭제할 수 없다.

### 지하철 노선 관리 기능

- [x] 지하철 노선을 등록할 수 있다.
  - [x] 지하철 노선의 이름은 2글자 이상이어야 한다.
  - [x] 중복된 지하철 노선 이름이 등록될 수 없다.
  - [x] 지하철 노선의 이름은 최대 10자 이하이다.
  - [x] 지하철 노선 생성시 필요한 값은 색상, 상행역, 하행역, (최초 상행역과 하행역 구간의)거리, 시간이다.
- [x] 지하철 노선을 수정할 수 있다.
  - [x] 지하철 노선의 이름, 상행역, 하행역, 거리, 시간, 색상을 수정할 수 있다.
- [x] 지하철 노선을 삭제할 수 있다.
  - [x] 삭제 시 confirm을 이용하여 한 번 더 유저가 확인할 수 있어야 한다.

### 지하철 구간 관리 기능

- [ ] 지하철 노선에 구간을 추가, 수정, 삭제할 수 있다.
- [ ] 구간 조회 시, 노선을 변경하면 해당 노선의 Color로 Select 엘리먼트의 배경색을 변경한다.

### 회원 기능

- [x] 유저는 회원 가입을 할 수 있다,
  - [x] 회원 가입시 받는 정보는 `email`, `name`, `password`이다.
- [x] 유저는 로그인 할 수 있다.
  - [x] 로그인하고 나면 로그인 버튼은 로그아웃 버튼으로 변경되어야 한다.
  - [x] 로그인한 유저만, 정보의 수정이 가능하다.
  - [x] 로그인하지 않은 유저에게는 로그인 외 다른 버튼은 보이지 않는다.
- [x] 유저는 로그아웃할 수 있다.
  - [x] 로그아웃하면 `/` 루트 페이지로 돌아온다.

<br>

moonheekim0118 committed:
init: 초기 설정 완료

moonheekim0118 committed:
chore : 바벨 설정 완료

moonheekim0118 committed:
chore : TS config 수정

moonheekim0118 committed:
chore : html 파일 자바스크립트 모듈 적용

moonheekim0118 committed:
chore : 패키지 description 수정

moonheekim0118 committed:
chore : 스크립트 수정

moonheekim0118 committed:
fix : 엔트리 파일 수정

moonheekim0118 committed:
chore : TSconfig 수정

moonheekim0118 committed:
feat : 라우터, 로그인페이지 구현

moonheekim0118 committed:
feat : 헤더 컴포넌트 구현

moonheekim0118 committed:
chore : TSconfig 수정

moonheekim0118 committed:
refactor : 라우터 추상화

moonheekim0118 committed:
feat : 초기 렌더링 구현
- Station -> Login 페이지로 넘어가는 기능 구현

moonheekim0118 committed:
feat : 하이퍼링크로 url 이동하도록 도와주는 함수 구현

moonheekim0118 committed:
feat : 스토어 객체 및 로그인 정보 스토어 생성

moonheekim0118 committed:
feat : 각 페이지에 AuthStore 적용

moonheekim0118 committed:
fix : 이벤트 바인딩 방식 수정

moonheekim0118 committed:
feat : 로그인 기능 구현 완료
- 로컬스토리지 추상화
- 로그인 서비스 구현 완료

moonheekim0118 committed:
feat : 회원가입 기능 구현

moonheekim0118 committed:
feat : 로그아웃 기능 구현

moonheekim0118 committed:
feat : 로그인 상태에 따라 헤더 뷰 변경

moonheekim0118 committed:
refactor : 렌더링 방식 수정
- 페이지와 컴포넌트 분리

moonheekim0118 committed:
feat : 회원가입 폼 컴포넌트 분리

moonheekim0118 committed:
feat : 역 리스트 컴포넌트 구현

moonheekim0118 committed:
refactor : 돔 생성 initDom으로 이동, bindEvent 접근제한자 수정

moonheekim0118 committed:
feat : 역 삭제 구현

moonheekim0118 committed:
fix : 역 이름 validator 추가

moonheekim0118 committed:
refactor: 코드 정리

moonheekim0118 committed:
chore : 바벨 설정 수정

moonheekim0118 committed:
feat : 역 이름 수정 구현

moonheekim0118 committed:
feat : 오버레이/버튼 클릭시 모달 닫기 구현

moonheekim0118 committed:
fix : 이벤트 바인딩 위치 수정

moonheekim0118 committed:
fix : 헤더 url 수정

moonheekim0118 committed:
refactor : 렌더 함수 통합
- 자식 컴포넌트와 page 컴포넌트에서 렌더함수 동일하게 작동하도록 수정

moonheekim0118 committed:
chore : 불필요한 파일 제거

moonheekim0118 committed:
refactor : 필요없는 argumenet 삭제

moonheekim0118 committed:
feat : 지하철 노선 등록 구현

moonheekim0118 committed:
feat : 노선으로 등록된 역 삭제 불가능하도록 수정

moonheekim0118 committed:
feat : 노선 삭제 구현

moonheekim0118 committed:
chore : tsconfig 에러 수정

moonheekim0118 committed:
refactor : props 전달방식 수정

moonheekim0118 committed:
feat : 노선 수정 기능 구현

moonheekim0118 committed:
fix : 노선 업데이트 id 값 관련 에러 수정

moonheekim0118 committed:
chore : 구간 관련 임시 삭제

moonheekim0118 committed:
feat : 404 페이지 추가

moonheekim0118 committed:
chore : 바벨 웹팩 설정 수정

moonheekim0118 committed:
chore : 웹팩 설정 수정

moonheekim0118 committed:
chore : 인덱스파일에 css import

moonheekim0118 committed:
feat : 베이스 url 수정

moonheekim0118 committed:
chore : 불필요한 파일 삭제

moonheekim0118 committed:
chore : 개행추가

moonheekim0118 committed:
style : 모달관련 css 수정

moonheekim0118 committed:
chore : 불필요한 바벨 플러그인 제거

HaJunRyu commented:
문희님! 유효성 검사 조건식에서 아마 <= 가 아니라 <로 작성된게 아닌가 싶습니다ㅎㅎ 배포하셔서 보기 쉽네요
![역 추가 기능](https://user-images.githubusercontent.com/71176945/121317320-26741400-c945-11eb-8f5a-8194b1646944.png)

HaJunRyu commented:
에공ㅠㅠ 이 경우에도 그렇네요 제가 구현할때 참고하려고 보다가 우연히 발견했습니다... 버그 찾으러 온게 아니었는데 버그 케이스는 충정로로 등록했다가 역삼으로 변경하고 새로 충정로를 등록한 다음 역삼 -> 충정로 노선을 찍었더니 같은걸로 인식하는것 같아서 상태가 바뀌지 않았나 싶습니다!
![노선관리 페이지](https://user-images.githubusercontent.com/71176945/121318240-085ae380-c946-11eb-8ee7-e8f3952738e1.png)

moonheekim0118 committed:
fix : 역 이름 validator 수정

moonheekim0118 committed:
fix : 역 이름 수정 시 중복 검사

moonheekim0118 committed:
fix : 역 이름 수정 시 alert 추가

moonheekim0118 committed:
fix : 역 이름 수정 시에도 validation 추가

ejinaaa commented on src/js/components/LineForm/index.ts:
```diff
@@ -0,0 +1,41 @@
+      lineService.add(
+        name.value,
+        upStation?.value ?? "",
+        downStation?.value ?? "",
```
|| 대신 null 병합 연산자를 사용하면 좌항에 들어오는 값을 더욱 확실히 할 수 있겠네요!

ejinaaa commented on src/js/components/LineList/view.ts:
```diff
@@ -0,0 +1,34 @@
+    ? `${lines
+        .map((line) => lineView(line.id, line.name, line.color))
+        .join("")}<hr class="my-0" />`
+    : "아직 등록된 노선이 없습니다.";
```
이런 메세지들은 enum으로 따로 묶어주면 어떨까요?

ejinaaa commented on src/js/components/LineUpdate/index.ts:
```diff
@@ -0,0 +1,61 @@
+        "#distance",
+        this.$container
+      ) as HTMLInputElement;
+      const { value: time } = $("#time", this.$container) as HTMLInputElement;
```
```{ value: time}```부분에서 time은 value의 타입이 되는건가요??

moonheekim0118 replied:
value 값을 time이란 변수명으로 빼오는거에용!

ejinaaa commented on src/js/components/StationUpdate/index.ts:
```diff
@@ -0,0 +1,34 @@
+  value: string;
+}
+
+class StationUpdate extends Component<IProps> {
```
```<IProps>```타입이름에 적혀있는 I는 무슨 의미인가요??

moonheekim0118 replied:
타입스크립트에서 Interface 네이밍을 저렇게 해주더라구요! 저도 이번에 처음 적용해봤습니다..ㅎㅎ

ejinaaa commented on src/js/constants/messages.ts:
```diff
@@ -0,0 +1,20 @@
+  NOT_CORRECT_LINE_STATION: "상행역과 하행역은 같을 수 없습니다.",
+  NOT_EXIST_LINE_STATION: "등록가능한 역이 없습니다.",
+};
+
```
요런것도 밑에 처럼 enum으로 만들면 어떨까요?

ejinaaa commented on src/js/core/localStorage.ts:
```diff
@@ -0,0 +1,48 @@
+  id: string;
+}
+
+class LocalStorage<T extends IDataBase> {
```
타입도 확장해서 쓸 수 있다는걸 자꾸 까먹게 되네요ㅜ

ejinaaa commented on src/js/core/router.ts:
```diff
@@ -0,0 +1,47 @@
+    } else {
+      notFound.render($main);
+    }
+  }
```
실행문이 하나인 경우 if, else의 중괄호를 생략해보면 어떨까요?

JunilHwang commented on .prettierrc:
```diff
@@ -0,0 +1,10 @@
+    "bracketSpacing": true,
+    "arrowParens": "always"
+  }
+
+  
```
흠.. 이 부분에 EOF 표기가 되고 있네요. 원인이 뭘까요?

ejinaaa reviewed (commented):
저는 변수명 하나 지을때도 주저리 주저리 늘어놓는데 문희씨 코드는 깔끔하고 읽기 편한 것 같아요! 오늘도 많이 배워갑니다!

JunilHwang commented on src/js/App.ts:
```diff
@@ -0,0 +1,20 @@
+  protected initChildren(): void {
+    const header = new Header();
+    const modal = new Modal();
+    authStore.addObserver(header);
+    uiStore.addObserver(modal);
```
옵저버 패턴을 잘 사용해주셨네요 👍 
조금더 자동화를 할 수 있는 방법이
굳이 observer를 통해서 등록하는게 아니라 Store를 Component에서 사용하기만 해도, 이를 자동으로 감지할 수 있는 방법은 없을까요?

```js
const authStore = observable({ a: 10, b: 20, c: 30 });
const observer1 = observe(() => console.log(`authStore.a = ${authStore.a}`));
const observer2 = observe(() => console.log(`authStore.b = ${authStore.b}`));
const observer3 = observe(() => console.log(`authStore.c = ${authStore.c}`));
authStore.a = 1; // observer1 호출
authStore.b = 2; // observer2 호출
authStore.c = 3; // observer3 호출
```

참고링크: https://meetup.toast.com/posts/188

moonheekim0118 replied:
오오 좋은 제안 감사드립니다!! 확실히 이렇게하면 관리하기 쉽겠네요!! 👍 👍

HaJunRyu replied:
링크 감사합니다 저도 잘 읽어보겠습니다!

JunilHwang commented on src/js/components/Header/index.ts:
```diff
@@ -0,0 +1,30 @@
+
+  protected componentMount(): void {
+    const { isLoggedIn } = authStore.getState();
+    this.$container.innerHTML = view(isLoggedIn);
```
view를 아예 Component에 포함시키면 어떨까요?
view와 Component는 애초에 강하게 결합되어 있는 코드라서 함께 관리하는게 유지보수할 때 더 좋은 것 같아요!

JunilHwang commented on src/js/components/LineForm/index.ts:
```diff
@@ -0,0 +1,41 @@
+  protected bindEvents(): void {
+    this.$container.addEventListener("submit", (e: Event) => {
+      e.preventDefault();
+      const name = $("#name", this.$container) as HTMLInputElement;
+      const upStation = $("#up", this.$container) as HTMLSelectElement;
+      const downStation = $("#down", this.$container) as HTMLSelectElement;
+      const color = $("#colors", this.$container) as HTMLInputElement;
+      const distance = $("#distance", this.$container) as HTMLInputElement;
+      const time = $("#time", this.$container) as HTMLInputElement;
```
아예 form에서 값을 받아오면 어떨까요? form과 `new FormData`를 이용해보세요!

moonheekim0118 replied:
FormData 처음 알아갑니다..!! 감사해용 👍 👍

JunilHwang commented on src/js/components/LineList/index.ts:
```diff
@@ -0,0 +1,43 @@
+  lineUpdate = new LineUpdate();
+
+  protected initDom(): void {
+    this.$container = createElement({
+      tag: "ul",
+      className: "mt-3 pl-0",
+    });
```
```suggestion
    this.$container = createElement(`<ul class="mt-3 pl-0"></ul>`);
```
이런 형태로도 만들 수 있도록 createElement를 정의해주면 어떨까요?

moonheekim0118 replied:
이렇게하니까 훨씬 깔끔하네용.. 👍 ..한번 구현해보겠습니다!!

JunilHwang commented on src/js/components/LineList/index.ts:
```diff
@@ -0,0 +1,43 @@
+        remove: () => lineService.remove(lineId),
+      };
+      actions[id]();
+    });
```
컴포넌트 내부에서 event delegation과 관련된 메소드를 만들어서 관리해주면 어떨까요?

JunilHwang commented on src/js/components/LineList/view.ts:
```diff
@@ -0,0 +1,34 @@
+const view = (lines: ILine[]): string => {
+  return lines.length > 0
+    ? `${lines
+        .map((line) => lineView(line.id, line.name, line.color))
```
```suggestion
        .map(lineView)
```
lineView가 line 객체를 받아서 처리할 수 있는 형태로 만들어주면 어떨까요?

JunilHwang commented on src/js/components/LineUpdate/index.ts:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  protected bindEvents(): void {
+    this.$container.addEventListener("submit", (e: Event) => {
```
```suggestion
    const { $container } = this;
    $container.addEventListener("submit", (e: Event) => {
```
전체적으로 $container를 사용중이라서 이렇게 this라도 없앨 수 있도록 해보면 좋을 것 같아요!

JunilHwang commented on src/js/components/LineUpdate/index.ts:
```diff
@@ -0,0 +1,61 @@
+          time,
+        },
+        this.props
+      );
```
이 부분도 form을 이용해서 만들어보세요!

JunilHwang commented on src/js/components/StationList/index.ts:
```diff
@@ -0,0 +1,44 @@
+import { $, createElement, closest } from "@/utils/dom";
+
+class StationList extends Component {
+  stationUpdate = new StationUpdate();
```
```suggestion
  private stationUpdate = new StationUpdate();
```
외부에서 사용되는 변수가 아니기 때문에 private으로 제한해주면 좋을 것 같아요!

JunilHwang commented on src/js/core/localStorage.ts:
```diff
@@ -0,0 +1,48 @@
+  id: string;
+}
+
+class LocalStorage<T extends IDataBase> {
```
sessionStorage도 localStorage와 똑같은 인터페이스를 사용한답니다.
아예 Storage를 추상화해보면 어떨까요?

moonheekim0118 replied:
하나만 생각하고 둘은 생각 못했네요ㅎㅎㅎ

JunilHwang commented on src/js/service/stationService.ts:
```diff
@@ -0,0 +1,64 @@
+    } catch (error) {
+      alert(error);
+    }
+  },
```
service 단에서 catch를 하기보단 app 단에서 catch를 해주면 어떨까요?
그러면 app에서 더 다양한 UI를 만들 수 있을 것 같아요!

JunilHwang commented on src/js/service/stationService.ts:
```diff
@@ -0,0 +1,64 @@
+  add: (name: string): void => {
+    try {
+      if (!stationNameValidator(name)) throw MESSAGE.NOT_CORRECT_STATION;
+      if (isExistedStation(name)) throw MESSAGE.EXIST_STATION;
```
throw를 단순하게 Message로 던지기보단, Error를 상속한 CustomError를 만들어서 관리해주면 어떨까요?
CustomError로 관리하게 될 경우 값으로 throw를 던지는 것 보다 다양한 방식으로 컨트롤을 할 수 있게 된답니다!

JunilHwang commented on src/js/service/stationService.ts:
```diff
@@ -0,0 +1,64 @@
+      if (!stationNameValidator(newName)) throw MESSAGE.NOT_CORRECT_STATION;
+      if (isExistedStation(newName)) throw MESSAGE.EXIST_STATION;
+      const newData = stationDB.update(id, { name: newName });
+      stationStore.updateState({ stations: newData });
+      uiService.closeModal();
```
Service와 Service 간에는 참조(의존)를 최소화 하면 어떨까요?
저같은 경우 Service끼리 섞이지 않게 했고, Service로직은 Store에서 일괄적으로 처리하도록 만들었어요 ㅎㅎ
그럴 경우 다른 Service를 신경 쓰지 않아도 되기 때문에 편하답니다!

JunilHwang reviewed (commented):
안녕하세요 문희님! 한 주 만에 코드가 더 읽기 수월해졌네요 👍
핵심적인 내용만 정리하자면 다음과 같습니다.

- Observer 패턴 자동화하기
  - 참고링크: https://meetup.toast.com/posts/188
- View와 Component를 결합해보면 어떨까요? 어차피 강하게 결합된 성격의 코드들이랍니다.
- form을 전송할 때 FormData를 이용해보세요!
- try / catch는 Service 밖에서 하거나 혹은 catch 후에 다시 throw를 해주면 어떨까요? 그래야 UI에서 처리하기가 더욱 수월하답니다!
- throw를 값으로 던지기 보단 Error 객체를 만들어서 던지면 어떨까요?
- Service 레이어 간의 의존을 줄이고 Service를 사용하는 레이어에 필요한 로직을 위임해주면 어떨까요?

전체적으로 코드 스타일이 통일되어서 이해하기가 무척 수월하네요! 인상 깊게 잘 봤습니다 🙇‍♂️

이 외에 문희님께 개인적으로 여쭤보고 싶은게 있어서 슬랙 메세지를 남겼었는데, DM은 안 보시는 것 같아서 다시 지웠습니다 ㅎㅎ; 혹시 괜찮으실 때 저에게 DM 주실 수 있을까요?

moonheekim0118 committed:
refactor : 자식 컴포넌트 접근제한자 private으로 수정

moonheekim0118 committed:
feat : observable 작성

moonheekim0118 committed:
refactor : 뷰->컴포넌트 내부로 이동

moonheekim0118 committed:
refactor : createElement 템플릿 구현

moonheekim0118 committed:
refactor : 폼 데이터 적용

moonheekim0118 committed:
refactor : 로컬스토리지 추상화 -> 스토리지로 수정

moonheekim0118 committed:
chore : 클래스명 수정

moonheekim0118 committed:
refactor : 이벤트위임 메서드 구현

moonheekim0118 committed:
refactor : 필요없는 파라미터 삭제

moonheekim0118 committed:
refactor : 에러 객체 생성
- 기존에 메시지만 던져주던 것을 에러 객체로 수정

moonheekim0118 committed:
refactor : try~catch 컴포넌트 핸들러로 이동

moonheekim0118 committed:
feat : Store 클래스 수정
- store 내부에서 observer 생성
- getState 사용 시, observer에 등록
- setState 사용 시, observer 실행

moonheekim0118 committed:
refactor : 컴포넌트 구조 1차 수정

moonheekim0118 committed:
chore : 언마운트 메서드 삭제

moonheekim0118 committed:
feat : 페이지 변경 시 언마운트 구현

moonheekim0118 committed:
refactor : 컴포넌트 setUp 수정
- useSelector 에서 스토어 관련 메서드 사용하도록 수정
- useSelector에서 store 관련 메서드 사용 될 시 관찰
